lib/EnzymeCore/src/EnzymeCore.jl

```diff
@@ -58,6 +58,17 @@ Mark a function argument `x` of [`autodiff`](@ref Enzyme.autodiff) as duplicated
 auto-differentiate in respect to such arguments, with `dx` acting as an
 accumulator for gradients (so ``\\partial f / \\partial x`` will be *added to*)
 `∂f_∂x`.
+
+# Note
+
+Enzyme forces `x` and `∂f_∂x` to have the same types when constructing `Duplicated`.
+This is not a mathematical requirement, but rather a guardrail to prevent user error.
+The memory locations of `x` and `∂f_∂x` are accessed in the same way by the differentiation code, so they should have the same data layout.
+Equality of types is an approximation of this condition.
+
+In some ways, type equality is too strict, but in other ways it is not strict enough.
```

Comment:
Could you add an explanation here as to why, in addition to the example? I can imagine people getting confused otherwise.

Also would you be able to add the example from the issue with the pointer offsets?

Reply:
I added a small explanation on the too strict / too lax thing.

As for the pointer offset example, I cannot add it because I do not understand it. I gave you the beginner version of this explanation, which is the one suited for people like me ;)